synapse/storage/databases/main/devices.py

```diff
@@ -1529,50 +1529,53 @@ async def add_device_change_to_streams(
         if not device_ids:
             return None
 
-        async with self._device_list_id_gen.get_next_mult(
-            len(device_ids)
-        ) as stream_ids:
-            await self.db_pool.runInteraction(
-                "add_device_change_to_stream",
-                self._add_device_change_to_stream_txn,
+        num_stream_ids = max(
+            len(device_ids),
+            len(room_ids) * len(device_ids),
+            len(hosts) * len(device_ids),
+        )
+
+        context = get_active_span_text_map()
+
+        def add_device_changes_txn(txn, stream_ids):
+            self._add_device_change_to_stream_txn(
+                txn,
                 user_id,
                 device_ids,
                 stream_ids,
             )
 
-        if not room_ids:
-            return stream_ids[-1]
+            if not room_ids:
+                return
```

Comment:
I'd be quite tempted to omit this (I believe `simple_insert_many_txn` handles an empty list correctly?)

failing that, some justification as to why we can bail out here might be good. "If the user isn't in any rooms, then clearly they also cannot share any rooms with remote users, so we can bail out now."